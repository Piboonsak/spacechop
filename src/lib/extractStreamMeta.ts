import { spawn } from 'duplex-child-process';
import { Stream } from 'stream';


export default async (stream: Stream): Promise<any> => new Promise((resolve, reject) => {
  const proc = spawn('magick', ['-', 'json:']);
  stream.pipe(proc);
  const buffer = [];
  proc.on('data', (chunk) => {
    buffer.push(chunk);
  });
  proc.on('error', (err) => reject(err));
  proc.on('end', () => {
    let data = Buffer.concat(buffer).toString();
    // fix issues with convert json.
    // replace -nan with NaN
    data = data.replace(/\-nan/g, 'null');
    try {
      const json = JSON.parse(data);
      const [{ image }] = json;
      resolve(image);
    } catch (err) {
      reject(err);
    }
  });
});
