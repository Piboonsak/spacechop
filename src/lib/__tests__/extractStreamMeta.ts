import fs from 'fs';
import path from 'path';
import extractStreamMeta from '../extractStreamMeta';

describe('Extract stream meta', () => {
  const assetsFolder = path.join(__dirname, './../../test/assets');
  const filename = path.join(assetsFolder, 'cat.jpg');
  const stream = fs.createReadStream(filename);

  it('should extract meta data from image', async () => {
    const meta = await extractStreamMeta(stream);
    // ignore file properties in extraction.
    delete meta.properties['date:create'];
    delete meta.properties['date:modify'];
    // ignore imagemagick properties
    delete meta.elapsedTime;
    delete meta.pixelsPerSecond;
    delete meta.userTime;
    expect(meta).toMatchSnapshot();
  });

  it('should throw when stream is broken', async () => {
    const empty = fs.createReadStream('/config.yml');
    expect(extractStreamMeta(empty)).rejects.toThrow();
  });
});
