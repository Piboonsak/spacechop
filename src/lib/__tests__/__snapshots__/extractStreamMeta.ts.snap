// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Extract stream meta should extract meta data from image 1`] = `
Object {
  "backgroundColor": "#FFFFFF",
  "baseDepth": 8,
  "baseName": "-",
  "borderColor": "#DFDFDF",
  "channelDepth": Object {
    "blue": 1,
    "green": 8,
    "red": 8,
  },
  "channelStatistics": Object {
    "Blue": Object {
      "entropy": 0.933503,
      "kurtosis": 0.0654779,
      "max": 255,
      "mean": 164.381,
      "min": 0,
      "skewness": -0.259904,
      "standardDeviation": 50.9994,
    },
    "Green": Object {
      "entropy": 0.938473,
      "kurtosis": -0.179475,
      "max": 255,
      "mean": 163.397,
      "min": 0,
      "skewness": -0.237488,
      "standardDeviation": 51.6582,
    },
    "Red": Object {
      "entropy": 0.942212,
      "kurtosis": -0.363768,
      "max": 255,
      "mean": 162.04,
      "min": 0,
      "skewness": -0.204228,
      "standardDeviation": 52.0268,
    },
  },
  "chromaticity": Object {
    "bluePrimary": Object {
      "x": 0.15,
      "y": 0.06,
    },
    "greenPrimary": Object {
      "x": 0.3,
      "y": 0.6,
    },
    "redPrimary": Object {
      "x": 0.64,
      "y": 0.33,
    },
    "whitePrimary": Object {
      "x": 0.3127,
      "y": 0.329,
    },
  },
  "class": "DirectClass",
  "colorspace": "sRGB",
  "compose": "Over",
  "compression": "JPEG",
  "depth": 8,
  "dispose": "Undefined",
  "endianess": "Undefined",
  "filesize": "64823B",
  "format": "JPEG",
  "formatDescription": "JPEG",
  "gamma": 0.454545,
  "geometry": Object {
    "height": 700,
    "width": 1000,
    "x": 0,
    "y": 0,
  },
  "imageStatistics": Object {
    "Overall": Object {
      "entropy": 0.938063,
      "kurtosis": -0.165411,
      "max": 255,
      "mean": 163.273,
      "min": 0,
      "skewness": -0.234124,
      "standardDeviation": 51.5721,
    },
  },
  "intensity": "Undefined",
  "interlace": "JPEG",
  "iterations": 0,
  "matteColor": "#BDBDBD",
  "mimeType": "image/jpeg",
  "name": null,
  "numberPixels": "700000",
  "orientation": "Undefined",
  "pageGeometry": Object {
    "height": 700,
    "width": 1000,
    "x": 0,
    "y": 0,
  },
  "pixels": 2100000,
  "properties": Object {
    "jpeg:colorspace": "2",
    "jpeg:sampling-factor": "2x2,1x1,1x1",
    "signature": "1c14fc6f10b34266cae55b6b6fde97ff0badfaa8f64745ec295046402e8c0ab1",
  },
  "quality": 85,
  "renderingIntent": "Perceptual",
  "tainted": false,
  "transparentColor": "#00000000",
  "type": "TrueColor",
  "units": "Undefined",
  "version": "file:///usr/share/doc/ImageMagick-7/index.html",
}
`;
